devlib/module/sched.py

```diff
@@ -0,0 +1,239 @@
+#    Copyright 2018 Arm Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+
+from devlib.module import Module
+from devlib.utils.misc import memoized
+
+class SchedModule(Module):
+
+    name = 'sched'
+
+    sysfs_root = '/sys/devices/system/'
+
+    @staticmethod
+    def probe(target):
+        return True
+
+    def get_sd_info(self, cpu):
+        return SchedDomainData(self.target, cpu)
+
+    def get_capacity(self, cpu):
+        """
+        :returns: The capacity of 'cpu'
+        """
+        return self.get_capacities()[cpu]
+
+    @memoized
+    def _has_em(self, cpu):
+        """
+        :returns: Whether energy model data is available for 'cpu'
+        """
+        procfs =  SchedDomainData(self.target, cpu).procfs
+        return procfs["domain0"].get("group0", {}).get("energy", {}).get("cap_states") != None
+
+    @memoized
+    def _has_dmips_capacity(self, cpu):
+        """
+        :returns: Whether dmips capacity data is available for 'cpu'
+        """
+        return self.target.file_exists(
+            os.path.join(self.sysfs_root, 'cpu/cpu{}/cpu_capacity'.format(cpu))
+        )
+
+    @memoized
+    def _get_em_capacity(self, cpu):
+        """
+        :returns: The maximum capacity value exposed by the EAS energy model
+        """
+        sd = SchedDomainData(self.target, cpu)
+
+        cap_states = sd.domains[0].groups[0].energy.cap_states
+        return int(cap_states.split('\t')[-2])
+
+    @memoized
+    def _get_dmips_capacity(self, cpu):
+        """
+        :returns: The capacity value generated from the capacity-dmips-mhz DT entry
+        """
+        return self.target.read_value(
+            os.path.join(
+                self.sysfs_root,
+                'cpu/cpu{}/cpu_capacity'.format(cpu)
+            ),
+            int
+        )
+
+    @memoized
+    def get_capacities(self):
+        """
+        :returns:  a dictionnary of the shape {cpu : capacity}
+        """
+        cpus = range(0, self.target.number_of_cpus)
+
+        capacities = {cpu : 0 for cpu in cpus}
+
+        for cpu in cpus:
+            if self._has_em(cpu):
+                capacities[cpu] = self._get_em_capacity(cpu)
+            elif self._has_dmips_capacity(cpu):
+                capacities[cpu] = self._get_dmips_capacity(cpu)
+            else:
+                self.logger.warning('Cannot read capacity for cpu{}, assuming 1024'.format(cpu))
+                capacities[cpu] = 1024
```

Comment:
Yeah, I'm unsure about this one. One the one hand it handles the case of returning 1024 for targets that don't expose any of those nodes (which is what we want), on the other it will not scream if those nodes are not present on a target where we *want* them to be here.